python-checks/src/test/resources/checks/specialMethodReturnTypeCheck.py

```diff
@@ -0,0 +1,319 @@
+from external import unknown
+
+def string_unknown() -> str:
+    ...
+
+def bool_unknown() -> bool:
+    ...
+
+class BoolMethodCheck01:
+    def __bool__(self):
+        return "Hello" # Noncompliant {{Return a value of type `bool` here.}}
+        #      ^^^^^^^
+
+class BoolMethodCheck02:
+    def __bool__(self):
+        return True # Compliant
+
+class BoolMethodCheck03:
+    def __bool__(self):
+        x = "Hello"
+        return x # Noncompliant
+
+class BoolMethodCheck04:
+    def __init__(self, condition):
+        self.condition = condition
+
+    def __bool__(self):
+        return self.condition # Compliant but potential for FNs: Technically, it is possible that this may not be a boolean but there is no explicit indication for it. Marking this may lead to FPs
+
+class BoolMethodCheck05:
+    def __init__(self, condition):
+        self.condition = condition
+
+    def __bool__(self):
+        return self.condition # FN: Requires better analysis tooling, see guaranteed invocation below
```

Comment:
Yes and at the same time, even with a better analysis, the primary location should not be here, but at the call site, i.e. at line 37, because that specific instance of `BoolMethodCheck05` is problematic. Of course, line 35 will still be part of the issue reporting, as a secondary location.
You'll see that we take this approach in DBD, where we do support inter-procedural analysis.